hdf5/test/mt_id_test.c

```diff
@@ -3331,94 +7475,440 @@ void * mt_test_fcn_1(void * _params)
+     *
+     * Cells in this array are set to TRUE if this thread is is responsible for 
+     * registering and de-registring the type managed by the instance of 
+     * id_type_t at the same index in types_array[].
+     */
+    j = 0;
+    type_index = params->types_start;
+    for ( i = 0; i < NUM_ID_TYPES; i++ ) {
+
+        if ( i < type_index ) {
+
+            types_assigned[i] = FALSE;
+
+        } else {
+
+            assert( i == type_index );
+
+            assert(0 == atomic_load(&(types_array[type_index].successful_registers)));
+            assert(0 == atomic_load(&(types_array[type_index].failed_registers)));
+
+            try_register_type(type_index, params->cs, params->ds, params->rpt_failures, 
+                              params->thread_id);
+
+            assert(1 == atomic_load(&(types_array[type_index].successful_registers)));
+            assert(0 == atomic_load(&(types_array[type_index].failed_registers)));
+
+            types_assigned[i] = TRUE;
+
+            j++;
+
+            if ( j < params->types_count ) {
+
+                type_index += params->types_stride;
 
-        fprintf(stderr, "mt_test_fcn_1:%d: -- point 4\n", params->thread_id);
+            } else {  /* no more types assigned to this thread */
+
+                type_index = NUM_ID_TYPES; 
+            }
+        }
     }
 
-    if ( display_op_stats ) {
+    H5E_BEGIN_TRY {
 
-        H5I_dump_nz_stats(stdout, "verify_objects()");
-        H5I_clear_stats();
-    }
+    assert( j == params->types_count );
 
-    params->err_cnt += inc_type_refs(params->types_start, params->types_count / 2, params->types_stride,
-                                     params->cs, params->ds, params->rpt_failures, params->thread_id);
+    /* setup the id progress array, and initialize it to zero */
+    if ( NULL == (id_op_counts = (int *)malloc(((size_t)(NUM_ID_INSTANCES)) * sizeof(int))) ) {
 
-    if ( show_progress ) {
+        proceed = FALSE;
 
-        fprintf(stderr, "mt_test_fcn_1:%d: -- point 5\n", params->thread_id);
+    } else {
+
+        for ( i = 0; i < NUM_ID_INSTANCES; i++ ) {
+ 
+            id_op_counts[i] = 0;
+        }
     }
 
-    if ( display_op_stats ) {
+    while ( ( proceed ) && ( ids_completed < params->ids_count ) ) {
 
-        H5I_dump_nz_stats(stdout, "inc_type_refs()");
-        H5I_clear_stats();
-    }
+        /* pick an ID */
+        
+        target_id_index = params->ids_start + ((rand() % params->ids_count) * params->ids_stride);
+        assert(target_id_index >= 0);
+        assert(target_id_index < NUM_ID_INSTANCES);
 
-    for ( i = 0; i < params->types_count; i++ ) {
+        /* load the target ID's assigned type index.  Note that this type is not guaranteed to 
+         * have benn registered yet unless it is one of the types assigned to this thread.
+         */
+        target_id_type_index = atomic_load(&(id_instance_array[target_id_index].type_index));
```

Comment:
Compiler on Mac OS complained about the data type of `type_index`.  Changing it to `Atomic int` solved the problem.

```
../../hdf5/test/mt_id_test.c:7686:32: error: address argument to atomic operation must be a pointer to _Atomic type ('int *' invalid)
        target_id_type_index = atomic_load(&(id_instance_array[target_id_index].type_index));
```

Reply:
John provided the fix by removing `atomic_load`.  It passed the test.